map/local_ads_helpers.cpp

```diff
@@ -0,0 +1,92 @@
+#include "map/local_ads_helpers.hpp"
+
+#include "coding/reader.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "base/assert.hpp"
+#include "base/bits.hpp"
+
+namespace
+{
+template<typename TPrimitive>
+TPrimitive ReadPrimitive(FileReader const & reader, uint64_t & offset)
+{
+  if (offset + sizeof(TPrimitive) > reader.Size())
+    MYTHROW(Reader::SizeException, ("File is corrupted", offset, sizeof(TPrimitive), reader.Size()));
+
+  std::vector<char> bytes(sizeof(TPrimitive));
+  ReadFromPos(reader, offset, bytes.data(), bytes.size());
+  offset += bytes.size();
+
+  MemReaderWithExceptions memReader(bytes.data(), bytes.size());
+  ReaderSource<decltype(memReader)> src(memReader);
+  return ReadPrimitiveFromSource<TPrimitive>(src);
+}
+}
+
+void WriteCountryName(FileWriter & writer, std::string const & countryName)
+{
+  ASSERT(!countryName.empty(), ());
+  ASSERT_LESS(countryName.length(), (1 << sizeof(uint8_t) * 8) - 1, ());
+  uint8_t len = static_cast<uint8_t>(countryName.length() + 1);
+  WriteToSink(writer, len);
+  writer.Write(countryName.c_str(), len);
+}
+
+void WriteDuration(FileWriter & writer, int64_t duration)
+{
+  uint64_t const encodedDuration = bits::ZigZagEncode(duration);
+  WriteToSink(writer, encodedDuration);
+}
+
+void WriteRawData(FileWriter & writer, std::vector<uint8_t> const & rawData)
+{
+  uint64_t const size = static_cast<uint64_t>(rawData.size());
+  WriteToSink(writer, size);
+  writer.Write(rawData.data(), size);
+}
+
+std::string ReadCountryName(FileReader const & reader, uint64_t & offset)
+{
+  uint8_t countryNameLen = 0;
+  if (offset + sizeof(countryNameLen) > reader.Size())
+    MYTHROW(Reader::SizeException, ("File is corrupted", offset, sizeof(countryNameLen), reader.Size()));
+
+  ReadFromPos(reader, offset, &countryNameLen, sizeof(countryNameLen));
+  offset += sizeof(countryNameLen);
+  ASSERT_NOT_EQUAL(countryNameLen, 0, ());
+
+  if (offset + countryNameLen > reader.Size())
+    MYTHROW(Reader::SizeException, ("File is corrupted", offset, countryNameLen, reader.Size()));
+
+  std::vector<char> bytes(countryNameLen);
+  ReadFromPos(reader, offset, bytes.data(), bytes.size());
+  offset += bytes.size();
+
+  if (bytes.back() != 0)
+    MYTHROW(Reader::ReadException, ("File is corrupted. Country name cannot be read"));
+
+  return std::string(static_cast<char const *>(bytes.data()));
```

Comment:
in that particular case 'bytes' should not contain \0 in the middle (because it is representation of a ZT-string), and bytes.back() is \0. if use std::string(bytes.begin(), bytes.end()) then string will contain '\0' in str and str.size() wont equal to strlen(str.c_str());
alternatively, ZT can be popped back from the 'bytes' and then can be used 'std::string(bytes.begin(), bytes.end())'. But then Roman's code looks clearer.

static_cast<const char*> is redundant here, bytes.data() is 'char*' and cast happens implicitly.